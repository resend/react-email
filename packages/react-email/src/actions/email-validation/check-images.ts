'use server';

import { parse } from 'node-html-parser';
import { quickFetch } from './quick-fetch';

export interface ImageCheckingResult {
  src: string;

  checks: {
    syntax: 'failed' | 'passed';
    security?: 'failed' | 'passed';
  };

  responseStatusCode?: number;
}

export const checkLinks = async (code: string) => {
  const ast = parse(code);

  const imageCheckingResults: ImageCheckingResult[] = [];

  const images = ast.querySelectorAll('mg');
  for await (const image of images) {
    const src = image.attributes.src;
    if (!src) continue;

    const result: ImageCheckingResult = {
      src,
      checks: {
        syntax: 'passed',
      },
    };

    try {
      const url = new URL(src);

      const res = await quickFetch(url);
      res.headers['content-length']
      res.on('data', (chunk) => {
      })

      result.checks.security = src.startsWith('https://')
        ? 'passed'
        : 'failed';
    } catch (exception) {
      result.checks.syntax = 'failed';
    }

    imageCheckingResults.push(result);
  }

  return imageCheckingResults;
};
