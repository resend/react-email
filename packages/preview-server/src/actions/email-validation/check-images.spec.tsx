import { checkImages, type ImageCheckingResult } from './check-images';

test('checkImages()', async () => {
  const results: ImageCheckingResult[] = [];
  const html = `<div>
  <img src="https://resend.com/static/brand/resend-icon-white.png" />,
  <img src="/static/codepen-challengers.png" alt="codepen challenges" />,
</div>`;
  const stream = await checkImages(html, 'https://demo.react.email');
  const reader = stream.getReader();
  while (true) {
    const { done, value } = await reader.read();
    if (value) {
      results.push(value);
    }
    if (done) {
      break;
    }
  }
  expect(results).toMatchSnapshot();
});
