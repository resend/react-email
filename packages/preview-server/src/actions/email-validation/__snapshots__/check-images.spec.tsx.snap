// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`checkImages() 1`] = `
[
  {
    "checks": [
      {
        "metadata": {
          "alt": undefined,
        },
        "passed": false,
        "type": "accessibility",
      },
      {
        "passed": true,
        "type": "syntax",
      },
      {
        "passed": true,
        "type": "security",
      },
      {
        "metadata": {
          "fetchStatusCode": 200,
        },
        "passed": true,
        "type": "fetch_attempt",
      },
      {
        "metadata": {
          "byteCount": 24534,
        },
        "passed": true,
        "type": "image_size",
      },
    ],
    "codeLocation": {
      "column": 3,
      "line": 2,
    },
    "source": "https://cdn.resend.com/brand/resend-icon-black.png",
    "status": "warning",
  },
  {
    "checks": [
      {
        "metadata": {
          "alt": "codepen challenges",
        },
        "passed": true,
        "type": "accessibility",
      },
      {
        "passed": true,
        "type": "syntax",
      },
      {
        "passed": true,
        "type": "security",
      },
      {
        "metadata": {
          "fetchStatusCode": 200,
        },
        "passed": true,
        "type": "fetch_attempt",
      },
      {
        "metadata": {
          "byteCount": 111922,
        },
        "passed": true,
        "type": "image_size",
      },
    ],
    "codeLocation": {
      "column": 3,
      "line": 3,
    },
    "source": "/static/codepen-challengers.png",
    "status": "success",
  },
]
`;
